### ISSUE TITLE
fix: CFBundleShortVersionString & CFBundleVersion Sync while Build Generation for iOS

### ISSUE THREAD

**Issue by tirth-multipl** [shorebirdtech/shorebird#1956]
**Description**

CFBundleShortVersionString & CFBundleVersion are not sync when generating iOS build with shorebird. These are soft-warnings. ⚠️ 

**Steps To Reproduce**

```bash
shorebird release ios --flavor prod  -- --dart-define=BASE_URL=<url>
```

**Expected Behavior**

CFBundleShortVersionString & CFBundleVersion should sync when generating iOS build with shorebird.

**Warnings**

```error
CFBundleShortVersionString Mismatch. The CFBundleShortVersionString value '1.15.32' of extension 'Runner.app/PlugIns/NotificationService.appex' does not match the CFBundleShortVersionString value '1.15.67' of its containing iOS application 'Runner.app'.

CFBundleVersion Mismatch. The CFBundleVersion value '2142' of extension 'Runner.app/PlugIns/NotificationService.appex' does not match the CFBundleVersion value '250' of its containing iOS application 'Runner.app'.
```

**Comment by bryanoltman** [comment]
If I'm understanding correctly, the issue is that your app extension version and your app version are not the same. Shorebird does not manipulate your app version or the versions of your app extensions. Are you suggesting that it should, or am I misunderstanding the problem?

**Comment by tirth-multipl** [comment]
yeah you're correct but when we do normal build generation w/o shorebird, this issue doesn't happen.

**Comment by bryanoltman** [comment]
When you say "normal build generation", what do you mean exactly? How do you normally perform your builds?

**Comment by tirth-multipl** [comment]
> When you say "normal build generation", what do you mean exactly? How do you normally perform your builds?

Without Shorebird I meant.

```
flutter build ipa --flavor prod --dart-define=BASE_URL=<url>
```

**Comment by eseidel** [comment]
I would be useful to understand which versions the Runner.app and Runner.appx end up with after a normal flutter build vs. a shorebird build.  It's not clear to me which version is correct.  I presume 1.15.67 is the "correct" version and the appx is just lagging behind for some reason?  I also wonder if `flutter` also isn't updating the version of the appx but swallowing the log?

`shorebird` just uses `flutter build` to do the underlying build, so it shouldn't affect this aspect.  Although it's possible that we're showing logs from `flutter build` that `flutter build` is hiding by default?

**Comment by bryanoltman** [comment]
I don't think we have enough information to understand and fix this issue. Closing due to inactivity.

**Comment by c0dezli** [comment]
Having the same problem, unable to release the app because of this

**Comment by eseidel** [comment]
> Having the same problem, unable to release the app because of this

Would love to help.  If you build with `flutter build ipa` does it work?

**Comment by tpkowastaken** [comment]
@c0dezli @tirth-multipl 
I got the same issue and found the solution.

So I use firebase messaging (fcm) for my push notifications. I need to also display small images inside of these notifications to better illustrate them. For that you need to add an extension to allow displaying images which is how the OP probably got here as well. Setup: [https://firebase.google.com/docs/cloud-messaging/flutter/receive#step_3_-_use_the_extension_helper](https://firebase.google.com/docs/cloud-messaging/flutter/receive#step_3_-_use_the_extension_helper)


Since this is a different target than Runner. It doesn't have marketing version and current project version set by flutter:

![Screenshot 2024-09-06 at 23 51 44](https://github.com/user-attachments/assets/f196356a-cbfb-4c4f-8122-68708ef1e934)

For OP it would be `2142` for the current project version and `1.15.32` for the marketing version.

Now this usually isn't a problem because `manageAppVersionAndBuildNumber` exists. This automatically sets the Runner version for all the targets.

However shorebird sets this to false. It apparently requires it. When I tried to use export_options.plist with it enabled it failed.

There is a way to sync the flutter version tho:

## 1. Create a debug and release config file

create two new config files in xcode. Name them something like short extension name and debug/release

![Screenshot 2024-09-06 at 23 59 25](https://github.com/user-attachments/assets/12c6e7fa-3965-4cbb-a643-f72f48458c49)
![Screenshot 2024-09-06 at 23 59 06](https://github.com/user-attachments/assets/b020d20c-4e89-48ed-805b-a3e778114423)
![Screenshot 2024-09-07 at 0 00 50](https://github.com/user-attachments/assets/c4de2c4f-8ecd-4af9-80db-4d49b1450e91)

## 2. Include the flutter generated config to those configs

debug config:
```
#include? "Pods/Target Support Files/Pods-NotificationService/Pods-NotificationService.debug.xcconfig"
#include "Flutter/Generated.xcconfig"
```

release config:
```
#include? "Pods/Target Support Files/Pods-NotificationService/Pods-NotificationService.release.xcconfig"
#include "Flutter/Generated.xcconfig"
```

### replace the `NotificationService` with your target name

## 3. Apply this config in the project info

![Screenshot 2024-09-07 at 0 08 11](https://github.com/user-attachments/assets/db484175-8c51-4d7b-a6a3-82a932d192b1)

(imgdebug in the image is the config file we just created for debug)

## 4. Set the variables to marketing version and current project version:

Current project version: `$(FLUTTER_BUILD_NUMBER)`
Marketing version: `$(FLUTTER_BUILD_NAME)`

![Screenshot 2024-09-07 at 0 10 14](https://github.com/user-attachments/assets/35ec8126-0cb7-40a8-902a-c40d7fb1df7b)

![Screenshot 2024-09-07 at 0 10 21](https://github.com/user-attachments/assets/f4656bd7-f5df-4f5e-be45-8b8ebf34c975)

If you managed to follow it correctly it should appear as the number.

## 5. Set it in the Notification target (or the one causing problems) info.plist as well:

```
	<key>CFBundleShortVersionString</key>
	<string>$(FLUTTER_BUILD_NAME)</string>
	<key>CFBundleVersion</key>
	<string>$(FLUTTER_BUILD_NUMBER)</string>
```

That should fix it! Good luck

**Comment by AhmedLSayed9** [comment]
This issue occurs to us when building using shorebird with `Notification Service Extension` on iOS:
<img width="500" alt="Screenshot 2024-10-22 at 7 24 09 AM" src="https://github.com/user-attachments/assets/de225238-c38a-4836-8b7e-0342ba186fb5">


The solution from @tpkowastaken didn't help (we're using flavors).

@bryanoltman, Can we re-open this?

**Comment by eseidel** [comment]
Happy to.

My current understanding is that Shorebird disables manageAppVersionAndBuildNumber.  I'm not sure I remember why we do this.

Shorebird needs to be able to read the exact app version of your app so that we can look up patches for it when requests come in from your app.  Your app will send requests to Shorebird's servers such as "{app_id: 1234-abcd, version: "1.0.0+1"}". When you build your app, we record the release number that it had at time of building.  I'm not sure if manageAppVersionAndBuildNumber was delaying this number setting until submission time or what.

But it sounds like manageAppVersionAndBuildNumber does more than we understood it to.  It doesn't just auto-manage the app version, it also manages the extension versions.  Since Flutter doesn't set the extension versions itself, when you have manageAppVersionAndBuildNumber off and build with Flutter (which is what Shorebird does under the covers) an extension version won't get set.

@bryanoltman do you remember why we set manageAppVersionAndBuildNumber to false?

**Comment by bryanoltman** [comment]
manageAppVersionAndBuildNumber causes Xcode to change the build number of the app after it has been packaged, meaning the version reported to shorebird the release created will be different than what the app reports once it has been released.

**Comment by eseidel** [comment]
Yeah, so I think our possible actions here are:
1. Change `shorebird` to read the release information from the ipa/archive *after* the adjustments by XCode are made.  It's unclear how to do that, however, since it's not clear when XCode makes the adjustments.  We could make `shorebird` do the archive step which presumably includes the version bump, but that's more complexity we'd need to explain to the user.
2. Remove Shorebird's use of the version code at all.  This is also needed for something like https://github.com/shorebirdtech/samples/issues/11 (which causes a bunch of problems for AAR users).  The difficulty there is that we use the version code right now as a short-hand for the uniqueness of the apk/ipa when asking our servers for updates.  The nice thing about the version code is that it's guaranteed unique by the stores.  If we generated our own UUID per release (for example) we would then make the build otherwise non-deterministic (since if you built the same sources twice they would get to different UUIDs).  One solution is that we could use the build UUID (that XCode generates) instead of the version-code on iOS?  I suspect that could work.  Hmm, but I think that's path-sensitive.  I believe XCode generates them on build and then *caches* the generation on a per-path basis.  So they're also non-deterministic, but cached per machine.

**Comment by eseidel** [comment]
A 3rd option would be to teach `flutter build ipa` to update the extension version codes, like it already knows how to set the app version codes.  I suspect that's actually the *easiest* option.

**Comment by alegos27** [comment]
Same issue here, are there any update o temporary solution?

**Comment by eseidel** [comment]
@bryanoltman do you agree that "option 3" is the correct path forward here?  e.g. that I should file a bug with `flutter` to have it set the version code of the extensions when it does for the app?

**Comment by bryanoltman** [comment]
I have not been able to reproduce this issue, so I'm not totally sure.

**Comment by bryanoltman** [comment]
Which isn't to say there's not an issue here, but I attempted to reproduce by adding a notification extension to the Flutter counter app example and going through the shorebird release+patch flow, and I did not see this happen, so I think there's more to this than simply "the app has an extension"

**Comment by AhmedLSayed9** [comment]
> Which isn't to say there's not an issue here, but I attempted to reproduce by adding a notification extension to the Flutter counter app example and going through the shorebird release+patch flow, and I did not see this happen, so I think there's more to this than simply "the app has an extension"

Can you check if you get the warnings when uploading to transporter?

**Comment by bryanoltman** [comment]
@AhmedLSayed9 happy to—I'm not sure I went that far before. Does everything work as intended up to that point? The original bug makes it seem like this should happen much earlier (when running `shorebird release ios`)

**Comment by AhmedLSayed9** [comment]
> @AhmedLSayed9 happy to—I'm not sure I went that far before. Does everything work as intended up to that point? The original bug makes it seem like this should happen much earlier (when running `shorebird release ios`)

I'm not sure but I get an email from Apple and warnings at transporter. I guess I have to test this again.
Btw, the counter example should have a different build version rather than the default `1.0.0+1` to re-produce this.

**Comment by bryanoltman** [comment]
Ok, I'm seeing the issue upon uploading using Transporter. Thanks! I'll look into it.

**Comment by eseidel** [comment]
It's possible running local validation would also trigger this?
https://stackoverflow.com/questions/7568420/how-to-perform-ios-app-validation-from-the-command-line

xcrun -sdk iphoneos Validation /path/to/App.{app or ipa}

**Comment by bryanoltman** [comment]
Error text, received via an email from Apple:

```
ITMS-90473: CFBundleShortVersionString Mismatch - The CFBundleShortVersionString value '1.0' of extension 'Runner.app/PlugIns/NotificationService.appex' does not match the CFBundleShortVersionString value '1.0.0' of its containing iOS application 'Runner.app'.
```

**Comment by AhmedLSayed9** [comment]
> Error text, received via an email from Apple:
> 
> ```
> ITMS-90473: CFBundleShortVersionString Mismatch - The CFBundleShortVersionString value '1.0' of extension 'Runner.app/PlugIns/NotificationService.appex' does not match the CFBundleShortVersionString value '1.0.0' of its containing iOS application 'Runner.app'.
> ```

Exactly.
Also, You should receive another warning `ITMS-90473: CFBundleVersion Mismatch` if you update the CFBundleVersion value of the app.

**Comment by bryanoltman** [comment]
Weirdly, this does not seem to be an issue when uploading a build created with `flutter build ipa --release`.

Edit: trying now with our ExportOptions.plist

**Comment by bryanoltman** [comment]
Confirming that this happens with stock Flutter when passed an ExportOptions.plist generated by Shorebird.

**Comment by bryanoltman** [comment]
Fixed by https://github.com/shorebirdtech/shorebird/pull/2620

**Comment by bryanoltman** [comment]
The fix for this was just released with Shorebird v1.4.12

**Comment by tpkowastaken** [comment]
Love your work. Thank you!

**Comment by bryanoltman** [comment]
Getting reports on Discord that this has not been fixed, reopening

**Comment by jacekendrick** [comment]
I am still getting this issue while using Flutter 3.29.3 and the Fastlane plugin for Shorebird (fastlane-plugin-shorebird (0.4.0))

**Comment by bryanoltman** [comment]
@jacekendrick can you share your fastfile?

**Comment by jacekendrick** [comment]
> [@jacekendrick](https://github.com/jacekendrick) can you share your fastfile?

Here it is! @bryanoltman 

```ruby
default_platform(:ios)

platform :ios do
  desc "Push a new beta build to TestFlight"

  before_all do
    xcode_select("/Applications/Xcode.app")
  end

  lane :beta do
    if is_ci
      create_keychain(
        name: ENV['MATCH_KEYCHAIN_NAME'],
        password: ENV["MATCH_KEYCHAIN_PASSWORD"],
        default_keychain: true,
        unlock: true,
        timeout: 3600,
        lock_when_sleeps: false
      )
    end

    match(
      type: "appstore",
      readonly: is_ci,
      keychain_name: ENV['MATCH_KEYCHAIN_NAME'],
      keychain_password: ENV["MATCH_KEYCHAIN_PASSWORD"],
    )

    flavor = ENV["FLAVOR"]
    bundle_id = ENV["BUNDLE_ID"]
    widget_identifier = bundle_id + ".RedFlagWidget"
    ipa_path = "../build/ios/ipa/Relay.ipa"

    increment_version_number(
      version_number: ENV["VERSION_NUMBER"],
    )

    increment_build_number(
      build_number: ENV["BUILD_NUMBER"],
    )

    begin
      shorebird_release(
        platform: "ios",
        args: "--flavor #{flavor} --flutter-version 3.29.3 --target lib/main_#{flavor}.dart",
      )
    rescue => e
      sh "echo '❌ Shorebird release failed with error: #{e.message}'"
      sh "echo 'Full error details:'"
      sh "echo '#{e.backtrace.join('\n')}'"
      raise e
    end

    api_key = app_store_connect_api_key(
      key_id: ENV["APPLE_API_KEY_ID"],
      issuer_id: ENV["APPLE_API_ISSUER_ID"],
      key_filepath: ENV["APPLE_API_P8_KEY_FILEPATH"],
      duration: 1200,
      in_house: false,
    )

    

    upload_to_testflight(
      ipa: ipa_path,
      skip_waiting_for_build_processing: true,
      api_key: api_key,
    )
  end
end
```

**Comment by Memadaldin** [comment]
@eseidel I'm getting the same error today.

ITMS-90473: CFBundleVersion Mismatch - The CFBundleVersion value '1' of extension 'Runner.app/PlugIns/CleverTapNotificationServiceExtension.appex' does not match the CFBundleVersion value '1719410970' of its containing iOS application 'Runner.app'.

ITMS-90473: CFBundleShortVersionString Mismatch - The CFBundleShortVersionString value '1.0' of extension 'Runner.app/PlugIns/CleverTapNotificationServiceExtension.appex' does not match the CFBundleShortVersionString value '3.50.0' of its containing iOS application 'Runner.app'.

ITMS-90473: CFBundleVersion Mismatch - The CFBundleVersion value '7' of extension 'Runner.app/PlugIns/OneSignalNotificationServiceExtension.appex' does not match the CFBundleVersion value '1719410970' of its containing iOS application 'Runner.app'.

ITMS-90473: CFBundleShortVersionString Mismatch - The CFBundleShortVersionString value '3.0.1' of extension 'Runner.app/PlugIns/OneSignalNotificationServiceExtension.appex' does not match the CFBundleShortVersionString value '3.50.0' of its containing iOS application 'Runner.app'.

sharing the lane `release_with_shorebird` we use for shorebird release
please tell me if extra info is needed.

``` Ruby

desc "Release using Shorebird"
  lane :release_with_shorebird do
    setup_ci
    branch = git_branch
    puts "*** Starting iOS Shorebird release ***"
    
    # Setup code signing
    api_key = lane_context[SharedValues::APP_STORE_CONNECT_API_KEY]
    sync_code_signing(
      api_key: api_key,
      type: "appstore",
      readonly: true,
      app_identifier: ["com.tradeling.iosmobilebuyerapp","com.tradeling.iosmobilebuyerapp.OneSignalNotificationServiceExtension", "com.tradeling.iosmobilebuyerapp.CleverTapNotificationServiceExtension"]
    )
    
    flutter_version = get_flutter_version
    version_parts = flutter_version.split('+')
    build_number = version_parts[1].to_i
    version_name = version_parts[0]
    
    puts "*** Building Shorebird release for version #{version_name}+#{build_number} ***"
    
    # Use the Shorebird plugin to handle the release
    shorebird_release(
      platform: "ios",
      args: "--flutter-version=3.29.0"
    )
    
    # The shorebird_release action sets lane_context[SharedValues::IPA_OUTPUT_PATH]
    # to the most recent IPA file, so we can use that
    ipa_path = lane_context[SharedValues::IPA_OUTPUT_PATH]
    puts "*** IPA created at: #{ipa_path} ***"
    
    begin
      upload_to_testflight(
        api_key: api_key,
        ipa: ipa_path,
        build_number: build_number.to_s,
        skip_waiting_for_build_processing: true,
        changelog: "[#{branch}] Shorebird release #{version_name}+#{build_number}"
      )
    rescue => ex
      puts "*** Error during upload: #{ex.message} ***"
      if ex.message.include?('Found more than 1 matching build')
        puts "*** Build appears to be already uploaded. Continuing... ***"
      else
        raise ex
      end
    end
    
    puts "*** Shorebird iOS release completed ***"
  end
```

**Comment by eseidel** [comment]
Thanks for sharing!
